### ISSUE TITLE
BlueprintInput.check_trigger_amount() modifies inputs

### ISSUE THREAD

**Issue by tsalo** [physiopy/phys2bids#314]
## Expected Behavior

Inputs to `BlueprintInput` should remain unmodified outside of the class (i.e., any changes to its inputs should be done on copies, rather than the original objects).

## Actual Behavior

`BlueprintInput.check_trigger_amount()` modifies the original time series corresponding "time". I have not dug into all of the `BlueprintInput` code looking for this issue specifically, so I don't know if this is the only case where inputs are modified.

This is a minor bug, as users will not generally use their original time series after creating the `BlueprintInput` object, but it could lead to unexpected behaviors later on.

## Steps to Reproduce the Problem

Here is a reduced version of @smoia's test that shows the bug:

```python
import numpy as np
from phys2bids.physio_obj import BlueprintInput

test_time = np.array([0, 1, 1, 2, 3, 5, 8, 13])
test_time_backup = test_time.copy()

blueprint_in = BlueprintInput(
    timeseries=[
        test_time,
        np.array([0, 1, 0, 0, 0, 0, 0, 0]),
        np.array([1, 0, 0, 1, 0, 0, 1, 0])
    ],
    freq=[42.0, 3.14, 20.0],
    ch_name=['time', 'trigger', 'chocolate'],
    units=['s', 's', 'sweetness'],
    trigger_idx=1
)

# Modify the time series in the object
blueprint_in.check_trigger_amount(thr=0.9, num_timepoints_expected=1)
assert np.array_equal(test_time, test_time_backup), "Array has been modified"
```

## Specifications

    - Python version: 3.6.4
    - phys2bids version: Current master
    - Platform: OSX

## Possible solution

I think the easiest solution is to store copies of the input arrays at `BlueprintInput.__init__()`.

**Comment by tsalo** [comment]
I think it would be quite uncommon for users to encounter this bug. I'm more worried about leaving it unresolved and then having a function added later on that creates `BlueprintInput` objects from in-memory data, but also uses the original data for something else.

**Comment by eurunuela** [comment]
The reason I asked is cause if it's not critical, we could add it as a GFI for Brainhack Donostia and make sure this issue is closed by the end of the event.

I think it's a perfect GFI for the event to be honest, and there's no need for a Python ninja to program it 😉

**Comment by vinferrer** [comment]
Guys, I just tried the code in my computer and the output doesn't get modified. Did someone address this without closing the issue?
```
import numpy as np 
   ...: from phys2bids.physio_obj import BlueprintInput 
   ...:  
   ...: test_time = np.array([0, 1, 1, 2, 3, 5, 8, 13]) 
   ...: test_time_backup = test_time.copy() 
   ...:  
   ...: blueprint_in = BlueprintInput( 
   ...:     timeseries=[ 
   ...:         test_time, 
   ...:         np.array([0, 1, 0, 0, 0, 0, 0, 0]), 
   ...:         np.array([1, 0, 0, 1, 0, 0, 1, 0]) 
   ...:     ], 
   ...:     freq=[42.0, 3.14, 20.0], 
   ...:     ch_name=['time', 'trigger', 'chocolate'], 
   ...:     units=['s', 's', 'sweetness'], 
   ...:     trigger_idx=1 
   ...: ) 
   ...:  
   ...: # Modify the time series in the object 
   ...: blueprint_in.check_trigger_amount(thr=0.9, num_timepoints_expected=1) 
   ...: assert np.array_equal(test_time, test_time_backup), "Array has been modified"                                                                                                                      

In [4]:                                                                                                                                                                                                    

In [4]: np.array_equal(test_time, test_time_backup),                                                                                                                                                       
Out[4]: (True,)
```

**Comment by vinferrer** [comment]
If not. Could someone tests this?

**Comment by vinferrer** [comment]
```
Python 3.8.5 (default, Jul 28 2020, 12:59:40) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.13.0 -- An enhanced Interactive Python. Type '?' for help.
```

**Comment by vinferrer** [comment]
i have been checking and the last line should be:
`assert np.array_equal(blueprint_in.timeseries[0], test_time_backup), "Array has been modified" `

**Comment by vinferrer** [comment]
```
        self.thr = thr
        self.time_offset = time_offset
        self.timeseries[0] = self.timeseries[0] - time_offset
        self.num_timepoints_found = num_timepoints_found
```
I think these are the lines that produces this change in `blueprint_in.timeseries[0]`

**Comment by vinferrer** [comment]
So which of these outputs do we want that change when executing `blueprint_in.check_trigger_amount`?

**Comment by vinferrer** [comment]
```
import numpy as np
from phys2bids.physio_obj import BlueprintInput

test_time = np.array([0, 1, 1, 2, 3, 5, 8, 13])
test_time_backup = test_time.copy()

blueprint_in = BlueprintInput(
    timeseries=[
        test_time,
        np.array([0, 1, 0, 0, 0, 0, 0, 0]),
        np.array([1, 0, 0, 1, 0, 0, 1, 0])
    ],
    freq=[42.0, 3.14, 20.0],
    ch_name=['time', 'trigger', 'chocolate'],
    units=['s', 's', 'sweetness'],
    trigger_idx=1
)
thr_backup=blueprint_in.thr
time_offset_backup=blueprint_in.time_offset
# Modify the time series in the object
blueprint_in.check_trigger_amount(thr=0.9, num_timepoints_expected=1)
np.array_equal(blueprint_in.timeseries[0], test_time_backup)

np.array_equal(blueprint_in.time_offset, time_offset_backup)

np.array_equal(blueprint_in.thr, thr_backup)
```
I have been doing trials, the thr does not appear in `blueprint_in` until the `check_trigger_amount` gets executed, 
timeseries[0] gets modified and  time_offset is also modified

**Comment by vinferrer** [comment]
```
        self.thr = thr
        self.time_offset = time_offset
        self.timeseries[0] = self.timeseries[0] - time_offset
        self.num_timepoints_found = num_timepoints_found
```
I propose all these become outputs to the function and we assign them outside the method

**Comment by vinferrer** [comment]
Has I explained in #357,  I couldn't reproduce the problem, if ``.copy()``  was removed  from the test_lines, (#358) tests still pass. I talked to @smoia  and now he cannot reproduce the bug in his computer and test pass also. Therefore I propose to remove the ``.copy`` as is done in #358 and be done with this issue

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.3.0`](https://github.com/physiopy/phys2bids/releases/tag/2.3.0) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.3.0`](https://github.com/physiopy/phys2bids/releases/tag/2.3.0) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.3.0`](https://github.com/physiopy/phys2bids/releases/tag/2.3.0) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.3.0`](https://github.com/physiopy/phys2bids/releases/tag/2.3.0) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.3.0`](https://github.com/physiopy/phys2bids/releases/tag/2.3.0) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.3.0`](https://github.com/physiopy/phys2bids/releases/tag/2.3.0) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.4.0`](https://github.com/physiopy/phys2bids/releases/tag/2.4.0) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.4.1`](https://github.com/physiopy/phys2bids/releases/tag/2.4.1) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.4.1`](https://github.com/physiopy/phys2bids/releases/tag/2.4.1) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.4.1`](https://github.com/physiopy/phys2bids/releases/tag/2.4.1) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.5.0`](https://github.com/physiopy/phys2bids/releases/tag/2.5.0) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.6.0`](https://github.com/physiopy/phys2bids/releases/tag/2.6.0) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.8.1`](https://github.com/physiopy/phys2bids/releases/tag/2.8.1) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.8.2`](https://github.com/physiopy/phys2bids/releases/tag/2.8.2) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.8.3`](https://github.com/physiopy/phys2bids/releases/tag/2.8.3) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.8.4`](https://github.com/physiopy/phys2bids/releases/tag/2.8.4) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.8.4`](https://github.com/physiopy/phys2bids/releases/tag/2.8.4) :rocket:

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.9.0`](https://github.com/physiopy/phys2bids/releases/tag/2.9.0) :rocket:
